Implement full search for text directives for Scroll to Text Fragment spec.

megangardner opened the pull request:
#### 721922be3ab62d8bdbde2ed93b6cb60d4cede316
<pre>
Implement full search for text directives for Scroll to Text Fragment spec.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243511">https://bugs.webkit.org/show_bug.cgi?id=243511</a>
rdar://89301977

Reviewed by Ryosuke Niwa.

Add full search algorithm as specified in the spec here:
<a href="https://wicg.github.io/scroll-to-text-fragment/">https://wicg.github.io/scroll-to-text-fragment/</a>
To properly search for text fragments from the directives that we have
parsed out of the URL.

Currently these have been tested against some examples, but in the
near future we will import that WPT tests and ensure that all of those pass
as well as address the last few edge case FIXMEs left in this implementation.

Source/WebCore/dom/FragmentDirectiveParser.cpp:
Fix an error in parsing the fragment directive that wasn&apos;t caught until I
finished coding the search algorithm.

* Source/WebCore/dom/BoundaryPoint.h:
(WebCore::operator!=):
* Source/WebCore/dom/FragmentDirectiveParser.cpp:
(WebCore::FragmentDirectiveParser::parseFragmentDirective):
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::isSearchInvisible):
(WebCore::FragmentDirectiveRangeFinder::isNonSearchableSubtree):
(WebCore::FragmentDirectiveRangeFinder::nearestBlockAncestor):
(WebCore::FragmentDirectiveRangeFinder::getBoundaryPointAtIndex):
(WebCore::FragmentDirectiveRangeFinder::indexIsWordBoundary):
(WebCore::FragmentDirectiveRangeFinder::isVisibleTextNode):
(WebCore::FragmentDirectiveRangeFinder::findRangeFromNodeList):
(WebCore::FragmentDirectiveRangeFinder::rangeOfStringInRange):
(WebCore::FragmentDirectiveRangeFinder::advanceRangeStartToNextNonWhitespace):
(WebCore::FragmentDirectiveRangeFinder::rangesForFragments):
(WebCore::FragmentDirectiveRangeFinder::rangeForFragment):
(WebCore::FragmentDirectiveRangeFinder::collapseIfRootsDiffer): Deleted.
(WebCore::FragmentDirectiveRangeFinder::rangeOfString): Deleted.
* Source/WebCore/dom/FragmentDirectiveRangeFinder.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::isDisplayBlockLevel const):

Canonical link: <a href="https://commits.webkit.org/253383@main">https://commits.webkit.org/253383@main</a>
</pre>

cdumez commented on Source/WebCore/dom/BoundaryPoint.h:
```diff
@@ -68,6 +68,11 @@ inline bool operator==(const BoundaryPoint& a, const BoundaryPoint& b)
 
+inline bool operator!=(const BoundaryPoint& a, const BoundaryPoint& b)
+{
+    return a.container.ptr() != b.container.ptr() || a.offset != b.offset;
```
To avoid code duplication, we usually do something like:
`return !(a == b);`

cdumez replied:
To avoid duplication, we often do something like this:
`return !(a == b);`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
 namespace FragmentDirectiveRangeFinder {
 
-static SimpleRange collapseIfRootsDiffer(SimpleRange&& range)
+static bool isSearchInvisible(Node& node)
```
Seems we could pass the node as `const Node&`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+static bool isNonSearchableSubtree(Node& node)
```
Can probably pass the node as a `const Node&`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
 {
-    if (target.isEmpty())
+    Node* traversingNode = &node;
+    while(traversingNode) {
```
Missing space after `while`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+{
+    Node* curNode = &node;
+    while (curNode)
+    {
```
Usually goes at the end of previous line.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+
+static Node& nearestBlockAncestor(Node& node)
+{
+    Node* curNode = &node;
```
In the spirit of WebKit coding style, we should use `currentNode` instead of `curNode`. We avoid abbreviations unless well known.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*> textNodeList, bool isEnd)
```
We shouldn't be passing a Vector by value.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+{
+    unsigned counted = 0;
+    
+    for (auto* curNode : textNodeList) {
```
`currentNode`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*> textNodeList, bool isEnd)
+{
+    unsigned counted = 0;
```
Can we find a better name than `counted`? I don't know this area of code very well but maybe `startIndex`? `counted` really doesn't tell me much about what it's counting and what for.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    return ubrk_isBoundary(it, index);
+}
+
+static bool isVisibleTextNode(Node& node)
```
Could probably be a `const Node&`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange searchRange, Vector<Node*> nodes, bool wordStartBounded, bool wordEndBounded)
```
We shouldn't be passing a Vector by value, same likely goes for SimpleRange.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+        else
+            matchIndex = potentialIndex;
+        
+        start = getBoundaryPointAtIndex(*matchIndex, nodes, false);
```
We should be using enum classes to avoid obscure boolean parameters like these. No idea what `false` means by just looking at the call site.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+            matchIndex = potentialIndex;
+        
+        start = getBoundaryPointAtIndex(*matchIndex, nodes, false);
+        end = getBoundaryPointAtIndex(*matchIndex + query.length(), nodes, true);
```
ditto.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    
+}
+
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange searchRange, bool wordStartBounded, bool wordEndBounded)
```
Should likely not pass the SimpleRange by value.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange searchRange, bool wordStartBounded, bool wordEndBounded)
+{
+    while (!searchRange.collapsed()) {
+        Node* curNode = &searchRange.startContainer();
```
`currentNode`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+                range.start = { *newStart, 0 };
+                continue;
+            }
+            else
```
usually goes at the end of the previous line.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+
+        auto string = node.textContent();
+
+        if (string.substringSharingImpl(offset,6) == "&nbsp;"_s)
```
Missing space after `,`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+            offset += 6;
 
-    return resultRange.collapsed() ? std::nullopt : std::make_optional(resultRange);
+        if (string.substringSharingImpl(offset,5) == "&nbsp"_s)
```
Missing space after `,`

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+            else
+                return;
+        }
+        else
```
Usually goes at the end of the previous line.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    
+    while (!searchRange.collapsed()) {
+        if (!parsedTextDirective.prefix.isNull()) {
+            auto prefixMatch = rangeOfStringInRange(parsedTextDirective.prefix, searchRange, true, false);
```
Should use enum classes to avoid those obscure booleans.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+            
+            ASSERT(matchRange->start.container->isTextNode(), "MatchRange start is not a Text Node");
+            
+            potentialMatch = rangeOfStringInRange(parsedTextDirective.textStart, matchRange.value(), false, mustEndAtWordBoundary);
```
Should use enum classes.

cdumez commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    
+    StringBuilder searchBufferBuilder;
+    for (auto* node : nodes) {
+        if (node)
```
Why this if check, I don't think you can ever put null in your Vector? Also, in the other place where you iterate over the vector, you don't null check.

As a side note, we should make sure that nothing can destroy the nodes while they in your vector since you're only holding raw pointers. It looks dangerous but hopefully safe here?

youssefsoli commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+    
+    // TODO: If the node serializes as void.
+    
+    if (is<HTMLIFrameElement>(node) || is<HTMLImageElement>(node) || is<HTMLMeterElement>(node) || is<HTMLObjectElement>(node) || is<HTMLProgressElement>(node) || is<HTMLStyleElement>(node) || is<HTMLScriptElement>(node) || is<HTMLVideoElement>(node) || is<HTMLAudioElement>(node))
```
This can be multi-lined for readability.

youssefsoli commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -26,51 +26,368 @@
+        }
+        
+        if (rangeEndSearchRange.collapsed()) {
+            ASSERT(!parsedTextDirective.textEnd.isNull());
```
Do you want this assertion to run if `rangeEndSearchRange.collapsed()` is true before we start looping?

megangardner replied:
This is matching the spec:

If rangeEndSearchRange is collapsed then:
Assert: parsedValues’s textEnd item is non-null
Return null

youssefsoli replied:
Right, but what if rangeEndSearchRange is collapsed before we begin looping? This code won't run.

youssefsoli replied:
My mistake, we are outside of the `while` loop not inside.

hortont424 commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,358 @@
+    if (*matchIndex + query.length() > searchBuffer.length() - endInset)
+        return std::nullopt;
+    
+    RELEASE_ASSERT_WITH_MESSAGE(start, "Scroll To Text Fragment: Start cannot be null");
```
RELEASE_ASSERT is very aggressive; do you think it's appropriate to kill the web process just because of a text fragment lookup failure? I think probably these should be normal asserts, or should just bail from lookup, or something... not this.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,355 @@
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange& searchRange, Vector<Node*>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
```
Pleas don't use a vector of Node*. Use WeakPtr or Ref/RefPtr whichever is appropriate.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,355 @@
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    // TODO: The computed value of its display property is none.
+    
+    // TODO: If the node serializes as void.
```
https://html.spec.whatwg.org/multipage/parsing.html#serializes-as-void

smfr commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<RefPtr<Node>>& textNodeList, BoundaryPointIsAtEnd isEnd)
```
`const Vector<RefPtr<Node>>& textNodeList`. Usually we don't use "get" in function names.

rniwa replied:
Maybe `boundaryPointAtIndexInNodes`?

rniwa replied:
Add a link to https://wicg.github.io/scroll-to-text-fragment/#get-boundary-point-at-index?

rniwa replied:
Can we also make this function take a Vector<Ref<Text>>&?

smfr commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange& searchRange, Vector<RefPtr<Node>>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
```
Can searchRange and nodes be `const`?

smfr commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+    
+}
+
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange& searchRange, WordBounded wordStartBounded, WordBounded wordEndBounded)
```
Can `searchRange` be `const` or is it intended to communicate out to the caller when it changes?

rniwa replied:
We should name this function `findStringInRange` like in the spec text.
"range of string in range" sounds way too confusing.

rniwa replied:
It most certainly need to be const here. Spec text (in general) passes in variables via copy.

smfr commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
     return ranges;
 }
 
+std::optional<SimpleRange> rangeForFragment(ParsedTextDirective parsedTextDirective, Document& document)
```
Should ParsedTextDirective be a const reference?

smfr reviewed (commented):
Does this change progress any tests? Would be nice to know that we have thorough test coverage for this code.

alanbaradlay commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+        if (!currentNode->isTextNode() && currentNode->renderer() && currentNode->renderer()->style().isDisplayBlockLevel())
+            return *currentNode;
+        currentNode = currentNode->parentInComposedTree();
+    }
```
what's the expected behavior with out-of-flow content where the nearest block level element is not necessarily the containing block?

alanbaradlay replied:
also how does this work in column context where you can have spanner elements across multiple columns?

megangardner replied:
I was following the spec as detailed here: https://wicg.github.io/scroll-to-text-fragment/#nearest-block-ancestor
It sounds there's cases that aren't covered?

rniwa commented on Source/WebCore/dom/FragmentDirectiveParser.cpp:
```diff
@@ -101,7 +101,7 @@ void FragmentDirectiveParser::parseFragmentDirective(StringView fragmentDirectiv
         if (tokens.first().endsWith('-') && tokens.first().length() > 1) {
             auto takenFirstToken = tokens.takeFirst();
-            if (auto prefix = WTF::URLParser::formURLDecode(StringView(takenFirstToken).left(takenFirstToken.length() - 2)))
+            if (auto prefix = WTF::URLParser::formURLDecode(StringView(takenFirstToken).left(takenFirstToken.length() - 1)))
```
Could you explain this change in the commit message?

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
 {
-    return &range.start.container->rootNode() == &range.end.container->rootNode()
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    // TODO: The computed value of its display property is none.
```
Use FIXME instead of TODO.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+    // TODO: The computed value of its display property is none.
+    
+    // TODO: If the node serializes as void.
+    
```
Could you add a URL to the spec?
https://wicg.github.io/scroll-to-text-fragment/#search-invisible

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+    while (traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentInComposedTree();
```
This isn't quite right. [Shadow-including ancestor](https://dom.spec.whatwg.org/#concept-shadow-including-ancestor) doesn't mean in the composed tree.
Specifically, a content assigned to a slot shouldn't be considered as non-searchable.
Use `parentOrShadowHostNode` instead.

rniwa commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2447,6 +2448,13 @@ inline bool RenderStyle::isDisplayRegionType() const
         || display() == DisplayType::ListItem;
 }
 
+inline bool RenderStyle::isDisplayBlockLevel() const
```
Maybe add a link to https://wicg.github.io/scroll-to-text-fragment/#has-block-level-display?

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+    return std::nullopt;
+}
+
+static void advanceRangeStartToNextNonWhitespace(SimpleRange& range)
```
Drop a link to https://wicg.github.io/scroll-to-text-fragment/#next-non-whitespace-position?

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+
+        auto string = node.textContent();
+
+        if (string.substringSharingImpl(offset, 6) == "&nbsp;"_s)
```
Why not just `string.startsWith("&nbsp;"_s))`?

megangardner replied:
Honestly, to match the spec text exactly. I originally advanced to offset and did startsWith, but it ended up making the code match the spec very difficult. Also by matching the text, I believe I found an issue with the spec, which I've added a comment about which should show up in the next update.

cdumez replied:
startsWith() would only be identical if offset was 0, which doesn't necessarily seem to be the case.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,369 @@
+        if (string.substringSharingImpl(offset, 6) == "&nbsp;"_s)
+            offset += 6;
+
+        if (string.substringSharingImpl(offset, 5) == "&nbsp"_s)
```
Ditto.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+            if (!prefixMatch)
+                return std::nullopt;
+            
+            auto newStart = makeBoundaryPoint(makeContainerOffsetPosition(prefixMatch->start).next());
```
`Position::next` does a lot more than finding the next boundary point.
Maybe add a FIXME here or explain we differ from the spec in that regard.

rniwa replied:
Nit: two spaces between "bit" and "more"

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+            
+            searchRange.start = newStart.value();
+            matchRange = makeSimpleRange(prefixMatch->end, searchRange.end);
+            advanceRangeStartToNextNonWhitespace(matchRange.value());
```
Here, we should assert that matchRange contains a SimpleRange.
Also, I'd prefer making advanceRangeStartToNextNonWhitespace returning a new Range instead of mutating the argument.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive
+std::optional<SimpleRange> rangeForFragment(const ParsedTextDirective parsedTextDirective, Document& document)
```
We should call this function `findRangeFromTextDirective` like the spec says.
"fragment" has a very specific meaning in this spec and the rest of web platform
so I think it's more confusing to call this function that.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#next-non-whitespace-position
+static void advanceRangeStartToNextNonWhitespace(SimpleRange& range)
```
Please make this function return a new range instead of mutating the one being passed in.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        }
+        
+        if (isNonSearchableSubtree(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
```
I don't think this is right. composed tree isn't same thing as shadow-including tree order.
I suggest we skip shadow DOM stuff and just call regular NodeTraversal::nextSkippingChildren here for now.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        }
+
+        if (!isVisibleTextNode(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
```
Ditto. Just call NodeTraversal::next for now with a FIXME.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        offset++;
+        
+        if (offset >= node.length()) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
```
Ditto. We should just use NodeTraversal::next for now.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+    while (currentNode) {
+        if (!currentNode->isTextNode() && currentNode->renderer() && currentNode->renderer()->style().isDisplayBlockLevel())
+            return *currentNode;
+        currentNode = currentNode->parentInComposedTree();
```
It's unclear whether we should be using flat/composed tree or shadow including tree.
Let's just call parentNode for now to be consistent with elsewhere.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        Node* currentNode = &searchRange.startContainer();
+        
+        if (isNonSearchableSubtree(*currentNode)) {
+            if (auto newStart = nextSkippingChildrenInComposedTreeIgnoringUserAgentShadow(*currentNode))
```
Again, composed tree isn't same thing as shadow-including tree order.
Let us just call NodeTraversal::nextSkippingChildren for now.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        if (isNonSearchableSubtree(*currentNode)) {
+            if (auto newStart = nextSkippingChildrenInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                searchRange.start = { *newStart, 0 };
+            else
```
Instead of adding else return everywhere.
Why don't we make `searchRange` `optional<SimpleRange>`?
And just check `searchRange && searchRange->collapsed()` as the loop invariant?

megangardner replied:
This only removed two returns and starts to deviate from the shape of the spec, I'd rather not.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        
+        if (!isVisibleTextNode(*currentNode)) {
+            do {
+                if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(*currentNode)) {
```
Ditto about using NodeTraversal::next for now.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange& searchRange, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    while (!searchRange.collapsed()) {
+        Node* currentNode = &searchRange.startContainer();
```
Please use RefPtr.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+                    currentNode = newStart;
+                } else
+                    return std::nullopt;
+            } while (is<DocumentType>(currentNode));
```
I think this will read better if we looped over DocType nodes and set range once instead.

megangardner replied:
I struggled with this sections a lot, I do believe it needs to be done as a do-while to match the spec.

rniwa replied:
ok

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+                break;
+
+            if (isSearchInvisible(*currentNode)) {
+                if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
```
This should be NodeTraversal::nextSkippingChildren.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+                    break;
+                continue;
+            }
+            if (currentNode->isTextNode())
```
The spec also checks the visibility of this text node here.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+            }
+            if (currentNode->isTextNode())
+                textNodeList.append(currentNode);
+            if (auto newNode = nextInComposedTreeIgnoringUserAgentShadow(*currentNode))
```
Ditto about using normal NodeTraversal::next here instead.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-node-list
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, const SimpleRange& searchRange, Vector<RefPtr<Node>>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
```
Since we're collecting a list of nodes we know are not null, we should take const Vector<Ref<Node>> here

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto node : nodes)
```
auto&

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+    StringBuilder searchBufferBuilder;
+    for (auto node : nodes)
+        searchBufferBuilder.append(downcast<Text>(*node).data());
+    searchBuffer = searchBufferBuilder.toString();
```
This is going to be one massive string...

rniwa replied:
Can we use `SearchBuffer` in `TextIterator.h` instead?
Maybe add a FIXME for now.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+        start = boundaryPointAtIndexInNodes(*matchIndex, nodes, BoundaryPointIsAtEnd::No);
+        end = boundaryPointAtIndexInNodes(*matchIndex + query.length(), nodes, BoundaryPointIsAtEnd::Yes);
+        
+        if (((wordStartBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex)) || ((wordEndBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex + query.length()))) {
```
We should probably call this function indexIs*At*WordBoundary

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+{
+    unsigned currentLength = 0;
+    
+    for (auto currentNode : textNodeList) {
```
auto&

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,381 @@
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#is-at-a-word-boundary
+static bool indexIsWordBoundary(const String& string, unsigned index)
```
Nit: This should be indexIs*At*WordBoundary

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,391 @@
+{
+    String searchBuffer;
+    std::optional<BoundaryPoint> start;
+    std::optional<BoundaryPoint> end;
```
Can we declare these two after setting searchStart to match the spec?

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,391 @@
+        start = boundaryPointAtIndexInNodes(*matchIndex, nodes, BoundaryPointIsAtEnd::No);
+        end = boundaryPointAtIndexInNodes(*matchIndex + query.length(), nodes, BoundaryPointIsAtEnd::Yes);
+        
+        if (((wordStartBounded == WordBounded::Yes) && !indexIsAtWordBoundary(searchBuffer, *matchIndex)) || ((wordEndBounded == WordBounded::Yes) && !indexIsAtWordBoundary(searchBuffer, *matchIndex + query.length()))) {
```
Nit: Can we wrap this line at `||`? This is awfully long if expression.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,391 @@
+        
+        auto& blockAncestor = nearestBlockAncestor(*currentNode);
+        Vector<Ref<Node>> textNodeList;
+        while (currentNode->isDescendantOf(blockAncestor) && is_lteq(treeOrder(BoundaryPoint(*currentNode, 0), searchRange.end))) {
```
Hm... this is O(n^2) since treeOrder will also do traversal.
Maybe add a FIXME?

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,391 @@
+                if (auto newNode = NodeTraversal::next(*currentNode))
+                    currentNode = newNode;
+                else
+                    break;
```
Instead of breaking like this, can we just add `!currentNode` to the loop invariant?

rniwa replied:
```
currentNode = NodeTraversal::next(*currentNode);
continue
```

rniwa reviewed (approved):
r=me after addressing the remaining comments.

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -28,49 +28,391 @@
+                break;
+
+            if (isSearchInvisible(*currentNode)) {
+                if (auto newNode = NodeTraversal::next(*currentNode))
```
I believe this should be NextTraversal::nextSkippingChildren to match the spec.
Since we want to skip over all the descendants.

megangardner committed:
Implement full search for text directives for Scroll to Text Fragment spec.
https://bugs.webkit.org/show_bug.cgi?id=243511
rdar://89301977

Reviewed by Ryosuke Niwa.

Add full search algorithm as specified in the spec here:
https://wicg.github.io/scroll-to-text-fragment/
To properly search for text fragments from the directives that we have
parsed out of the URL.

Currently these have been tested against some examples, but in the
near future we will import that WPT tests and ensure that all of those pass
as well as address the last few edge case FIXMEs left in this implementation.

Source/WebCore/dom/FragmentDirectiveParser.cpp:
Fix an error in parsing the fragment directive that wasn't caught until I
finished coding the search algorithm.

* Source/WebCore/dom/BoundaryPoint.h:
(WebCore::operator!=):
* Source/WebCore/dom/FragmentDirectiveParser.cpp:
(WebCore::FragmentDirectiveParser::parseFragmentDirective):
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::isSearchInvisible):
(WebCore::FragmentDirectiveRangeFinder::isNonSearchableSubtree):
(WebCore::FragmentDirectiveRangeFinder::nearestBlockAncestor):
(WebCore::FragmentDirectiveRangeFinder::getBoundaryPointAtIndex):
(WebCore::FragmentDirectiveRangeFinder::indexIsWordBoundary):
(WebCore::FragmentDirectiveRangeFinder::isVisibleTextNode):
(WebCore::FragmentDirectiveRangeFinder::findRangeFromNodeList):
(WebCore::FragmentDirectiveRangeFinder::rangeOfStringInRange):
(WebCore::FragmentDirectiveRangeFinder::advanceRangeStartToNextNonWhitespace):
(WebCore::FragmentDirectiveRangeFinder::rangesForFragments):
(WebCore::FragmentDirectiveRangeFinder::rangeForFragment):
(WebCore::FragmentDirectiveRangeFinder::collapseIfRootsDiffer): Deleted.
(WebCore::FragmentDirectiveRangeFinder::rangeOfString): Deleted.
* Source/WebCore/dom/FragmentDirectiveRangeFinder.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::isDisplayBlockLevel const):

Canonical link: https://commits.webkit.org/253383@main
